aptos-move/framework/src/natives/code.rs

```diff
@@ -124,24 +110,36 @@ pub struct PublishRequest {
     pub check_compat: bool,
 }
 
-/// Returns all natives for code module.
-pub fn code_natives(addr: AccountAddress) -> NativeFunctionTable {
-    native_functions::make_table(
-        addr,
-        &[
-            ("code", "request_publish", native_request_publish),
-            ("code", "from_bytes", native_from_bytes),
-        ],
-    )
+/// Gets the string value embedded in a Move `string::String` struct.
+fn get_move_string(v: Value) -> PartialVMResult<String> {
+    let bytes = v
+        .value_as::<Struct>()?
+        .unpack()?
+        .next()
+        .ok_or_else(|| PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR))?
+        .value_as::<Vec<u8>>()?;
+    String::from_utf8(bytes).map_err(|_| PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR))
+}
+
+/***************************************************************************************************
+ * native fun request_publish(
+ *     destination: address,
+ *     expected_modules: vector<String>,
+ *     code: vector<vector<u8>>,
+ *     policy: u8
+ * )
+ *
+ *   gas cost: base_cost + unit_cost * bytes_len
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct RequestPublishGasParameters {
+    pub base_cost: u64,
+    pub unit_cost: u64,
 }
```

Comment:
now we have a type that expects base and unit... it seems like we could have this as part of the native registration process and make it simpler...

fortunately I think this is something that can be iterated on if it becomes to cumbersome to wrap up in the gas only effort.

Reply:
I'd be cautious with requiring all native functions to have the `base_cost + unit_cost * len` model. It does work for the majority of them, but as long as there's one that doesn't, it will be a deal breaker. I'd also expect our gas formula to be more refined in the future, meaning that they will have more gas parameters.